tests/integrationv2/test_external_psk.py

```diff
@@ -0,0 +1,213 @@
+    random_bytes = data_bytes(64)
+    client_options = ProviderOptions(
+        mode=S2N.ClientMode,
+        host="localhost",
+        port=port,
+        cipher=Ciphers.CHACHA20_POLY1305_SHA256,
+        data_to_send=random_bytes,
+        key=Certificates.ECDSA_256.key,
+        cert=Certificates.ECDSA_256.cert,
+        trust_store=Certificates.ECDSA_256.cert,
+        insecure=False,
+        extra_flags=client_psk_params,
+        protocol=Protocols.TLS13)
+    
+    idx = S2N_CLIENT_PSK_PARAMETERS.index(client_psk_params)
+    server_psk_params = S2N_SERVER_PSK_PARAMETERS[idx]
+    server_options = copy.copy(client_options)
+    server_options.data_to_send = None
+    server_options.mode = S2N.ServerMode
+    server_options.extra_flags = server_psk_params
+
+    server = managed_process(S2N, server_options, timeout=20)
+    client = managed_process(S2N, client_options, timeout=20)
+
+    for results in client.get_results():
+        s2n_assert_chosen_psk(idx, results, random_bytes)
+        assert results.exception is None
+
+    for results in server.get_results():
+        s2n_assert_chosen_psk(idx, results, random_bytes, True)
+        assert results.exception is None
+
+pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("s2n_client_psk_params", S2N_CLIENT_PSK_PARAMETERS[:3], ids=get_parameter_name)
+def test_external_psk_s2nc_with_openssl_server(managed_process, s2n_client_psk_params):
+    port = next(available_ports)
+
+    random_bytes = data_bytes(64)
+    client_options = ProviderOptions(
+        mode=S2N.ClientMode,
+        host="localhost",
+        port=port,
+        cipher=Ciphers.CHACHA20_POLY1305_SHA256,
+        data_to_send=random_bytes,
+        insecure=False,
+        key=Certificates.ECDSA_256.key,
+        cert=Certificates.ECDSA_256.cert,
+        trust_store=Certificates.ECDSA_256.cert,
+        extra_flags=s2n_client_psk_params,
+        protocol=Protocols.TLS13)
+
+    server_options = copy.copy(client_options)
+    server_options.data_to_send = None
+    server_options.mode = OpenSSL.ServerMode
+    server_options.extra_flags = OPENSSL_PSK_PARAMETER
+    idx = S2N_CLIENT_PSK_PARAMETERS.index(s2n_client_psk_params)
+    server = managed_process(OpenSSL, server_options, timeout=20)
+    client = managed_process(S2N, client_options, timeout=20)
+
+    for results in client.get_results():
+        s2n_assert_chosen_psk(idx, results, random_bytes, False, True)
+        assert results.exception is None
+
+    for results in server.get_results():
+        if idx == 0:
+            assert b"PSK warning: client identity not what we expected" in results.stdout
+            assert b"error:141F906E:SSL routines:tls_parse_ctos_psk:bad extension:ssl/statem/extensions_srvr.c:1272" in results.stderr
+            assert results.exception is None
+            assert results.exit_code == 0
+        elif idx % 2 == 1:
+            assert b"PSK key given, setting server callback" in results.stdout
```

Comment:
So odd-indexed tests are always successful, and evens are failures? Can you configure your test cases in a more intuitive way?

Reply:
No there is more to it than just odd and even. I can make it more readable.